File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/transport/BLEConnector.kt
// BLEConnector.kt
// BlueSDK - BLE 连接器：管理 GATT 连接和数据收发

package com.blue.sdk.transport

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.os.Build
import com.blue.sdk.internal.BlueLogger
import java.util.UUID

internal interface BLEConnectorDelegate {
    fun onConnected()
    fun onDisconnected(error: Exception?)
    fun onDataReceived(data: ByteArray)
}

internal class BLEConnector {

    companion object {
        // LX-PD02 GATT 服务/特征 UUID
        private val SERVICE_UUID    = UUID.fromString("0000D459-0000-1000-8000-00805F9B34FB")
        private val WRITE_CHAR_UUID = UUID.fromString("00000013-0000-1000-8000-00805F9B34FB")
        private val NOTIFY_CHAR_UUID = UUID.fromString("00000014-0000-1000-8000-00805F9B34FB")
        private val CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805F9B34FB")
    }

    private var gatt: BluetoothGatt? = null
    private var writeCharacteristic: BluetoothGattCharacteristic? = null
    var delegate: BLEConnectorDelegate? = null

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            // 忽略非当前 GATT 实例的回调（防止旧连接残留回调导致重复处理）
            if (gatt != this@BLEConnector.gatt) return
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    BlueLogger.info("GATT connected, discovering services")
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    writeCharacteristic = null
                    val error = if (status != BluetoothGatt.GATT_SUCCESS)
                        Exception("GATT 断开，状态码：$status") else null
                    BlueLogger.info("GATT disconnected: ${error?.message ?: "normal"}")
                    delegate?.onDisconnected(error)
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (gatt != this@BLEConnector.gatt) return
            if (status != BluetoothGatt.GATT_SUCCESS) {
                BlueLogger.error("Service discovery failed, status: $status")
                delegate?.onDisconnected(Exception("服务发现失败"))
                return
            }
            val service = gatt.getService(SERVICE_UUID)
            if (service == null) {
                BlueLogger.error("Target service not found: $SERVICE_UUID")
                delegate?.onDisconnected(Exception("未找到目标服务"))
                return
            }
            writeCharacteristic = service.getCharacteristic(WRITE_CHAR_UUID)
            val notifyChar = service.getCharacteristic(NOTIFY_CHAR_UUID)
            if (notifyChar != null) {
                gatt.setCharacteristicNotification(notifyChar, true)
                val descriptor = notifyChar.getDescriptor(CCCD_UUID)
                if (descriptor != null) {
                    writeDescriptorCompat(gatt, descriptor, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
                }
            }
            if (writeCharacteristic != null) {
                BlueLogger.info("GATT characteristics ready, connected")
                delegate?.onConnected()
            }
        }

        @Suppress("DEPRECATION")
        @Deprecated("Use onCharacteristicChanged(gatt, characteristic, value) for API 33+")
        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            if (gatt != this@BLEConnector.gatt) return
            // API 33+ 上系统会同时触发新旧两个回调，旧回调的 characteristic.value 可能是过期数据
            // 仅在 API 32 及以下使用此回调
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) return
            val data = characteristic.value ?: return
            if (BlueLogger.rawFrameLogEnabled) {
                BlueLogger.debug("RX: ${data.joinToString(" ") { "%02X".format(it) }}")
            }
            delegate?.onDataReceived(data)
        }

        // API 33+ 的新回调
        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, value: ByteArray) {
            if (gatt != this@BLEConnector.gatt) return
            if (BlueLogger.rawFrameLogEnabled) {
                BlueLogger.debug("RX: ${value.joinToString(" ") { "%02X".format(it) }}")
            }
            delegate?.onDataReceived(value)
        }

        override fun onCharacteristicWrite(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                BlueLogger.error("Write failed, status: $status")
            }
        }
    }

    fun connect(context: Context, device: BluetoothDevice) {
        // 关闭旧的 GATT 连接（避免重复注册 notify 导致数据重复回调）
        gatt?.let {
            it.disconnect()
            it.close()
        }
        gatt = null
        writeCharacteristic = null
        BlueLogger.info("Connecting to: ${device.address}")
        gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    fun disconnect() {
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        writeCharacteristic = null
        BlueLogger.info("Disconnected manually")
    }

    fun write(bytes: ByteArray) {
        val char = writeCharacteristic
        val g = gatt
        if (char == null || g == null) {
            BlueLogger.error("Write characteristic not ready")
            return
        }
        writeCharacteristicCompat(g, char, bytes)
        if (BlueLogger.rawFrameLogEnabled) {
            BlueLogger.debug("TX: ${bytes.joinToString(" ") { "%02X".format(it) }}")
        }
    }

    // MARK: - API 版本兼容方法

    @Suppress("DEPRECATION")
    private fun writeCharacteristicCompat(
        gatt: BluetoothGatt,
        characteristic: BluetoothGattCharacteristic,
        value: ByteArray
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Android 13+ (API 33) 新版 API
            gatt.writeCharacteristic(characteristic, value, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT)
        } else {
            // Android 12 及以下旧版 API
            characteristic.value = value
            characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            gatt.writeCharacteristic(characteristic)
        }
    }

    @Suppress("DEPRECATION")
    private fun writeDescriptorCompat(
        gatt: BluetoothGatt,
        descriptor: BluetoothGattDescriptor,
        value: ByteArray
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Android 13+ (API 33) 新版 API
            gatt.writeDescriptor(descriptor, value)
        } else {
            // Android 12 及以下旧版 API
            descriptor.value = value
            gatt.writeDescriptor(descriptor)
        }
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Transport/BLEConnector.swift
// BLEConnector.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// BLE 连接器：管理与 LX-PD02 设备的 GATT 连接和数据收发
// 使用 BLECentralManager 单例，确保与扫描器共用同一个 CBCentralManager

import Foundation
import CoreBluetooth

/// BLE 连接器事件回调
protocol BLEConnectorDelegate: AnyObject {
    func bleConnectorDidConnect()
    func bleConnectorDidDisconnect(error: Error?)
    func bleConnectorDidReceiveData(_ data: Data)
}

/// BLE 连接器
/// 负责 GATT 连接建立、特征值读写和 Notify 订阅
final class BLEConnector: NSObject {

    // MARK: - GATT 服务/特征 UUID（LX-PD02）
    static let serviceUUID    = CBUUID(string: "D459")
    private static let writeCharUUID  = CBUUID(string: "0013") // 下行写特征
    private static let notifyCharUUID = CBUUID(string: "0014") // 上行通知特征

    // MARK: - 状态

    private let bleCentral = BLECentralManager.shared
    private var peripheral: CBPeripheral?
    private var writeCharacteristic: CBCharacteristic?
    private var notifyCharacteristic: CBCharacteristic?
    private let logger = BlueLogger.shared

    weak var delegate: BLEConnectorDelegate?

    // MARK: - 公开方法

    /// 连接指定设备
    func connect(peripheral: CBPeripheral) {
        self.peripheral = peripheral
        peripheral.delegate = self
        // 注册为连接事件代理，接收 centralManager 的连接/断开回调
        bleCentral.connectionDelegate = self
        bleCentral.centralManager.connect(peripheral, options: nil)
        logger.info("Connecting to device: \(peripheral.name ?? peripheral.identifier.uuidString)")
    }

    /// 断开连接
    func disconnect() {
        guard let peripheral = peripheral else { return }
        
        if let notifyChar = notifyCharacteristic {
            peripheral.setNotifyValue(false, for: notifyChar)
        }
        
        bleCentral.centralManager.cancelPeripheralConnection(peripheral)
        
        writeCharacteristic = nil
        notifyCharacteristic = nil
        self.peripheral = nil
        
        logger.info("Disconnected manually")
    }

    /// 发送数据帧
    func write(_ bytes: [UInt8]) {
        guard let peripheral = peripheral,
              let characteristic = writeCharacteristic else {
            logger.error("Write characteristic not ready")
            return
        }
        let data = Data(bytes)
        peripheral.writeValue(data, for: characteristic, type: .withResponse)
        if logger.rawFrameLogEnabled {
            logger.debug("TX: \(bytes.map { String(format: "%02X", $0) }.joined(separator: " "))")
        }
    }
}

// MARK: - BLECentralConnectionDelegate

extension BLEConnector: BLECentralConnectionDelegate {

    func centralDidConnect(peripheral: CBPeripheral) {
        logger.info("Device connected, discovering services")
        peripheral.discoverServices([BLEConnector.serviceUUID])
    }

    func centralDidDisconnect(peripheral: CBPeripheral, error: Error?) {
        writeCharacteristic = nil
        notifyCharacteristic = nil
        logger.info("Device disconnected: \(error?.localizedDescription ?? "normal")")
        delegate?.bleConnectorDidDisconnect(error: error)
    }

    func centralDidFailToConnect(peripheral: CBPeripheral, error: Error?) {
        logger.error("Connection failed: \(error?.localizedDescription ?? "unknown")")
        delegate?.bleConnectorDidDisconnect(error: error)
    }
}

// MARK: - CBPeripheralDelegate

extension BLEConnector: CBPeripheralDelegate {

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil,
              let service = peripheral.services?.first(where: { $0.uuid == BLEConnector.serviceUUID }) else {
            logger.error("Service discovery failed: \(error?.localizedDescription ?? "target service not found")")
            delegate?.bleConnectorDidDisconnect(error: error)
            return
        }
        peripheral.discoverCharacteristics(
            [BLEConnector.writeCharUUID, BLEConnector.notifyCharUUID],
            for: service
        )
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard error == nil else {
            logger.error("Characteristic discovery failed: \(error!.localizedDescription)")
            return
        }

        for characteristic in service.characteristics ?? [] {
            if characteristic.uuid == BLEConnector.writeCharUUID {
                writeCharacteristic = characteristic
            }
            if characteristic.uuid == BLEConnector.notifyCharUUID,
               characteristic.properties.contains(.notify) {
                notifyCharacteristic = characteristic
                peripheral.setNotifyValue(true, for: characteristic)
            }
        }

        if writeCharacteristic != nil {
            logger.info("GATT characteristics ready, connection complete")
            delegate?.bleConnectorDidConnect()
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard error == nil, let data = characteristic.value else { return }
        if logger.rawFrameLogEnabled {
            logger.debug("RX: \([UInt8](data).map { String(format: "%02X", $0) }.joined(separator: " "))")
        }
        delegate?.bleConnectorDidReceiveData(data)
    }

    func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        if let error = error {
            logger.error("Write failed: \(error.localizedDescription)")
        }
    }
}
